### PR TITLE
Added logic to deduplicate subscriptions

### PR DESCRIPTION
**Pull request by bobs4462**

### CODE REVIEW

**Comment by polachok** [comment]
Should be rebased on master

**Review by polachok — CHANGES_REQUESTED** [review]

File: src/pubsub.rs [hunk 1]
```diff
@@ -1281,6 +1347,14 @@ pub enum AccountCommand {
     Purge(Subscription, Commitment),
 }
 
+#[derive(Message, Debug)]
+#[rtype(result = "bool")]
+pub struct IsSubActive {
+    sub: Subscription,
+    commitment: Commitment,
+    owner: Option<Subscription>,
```
[the remark this answers is not in the document]
Comment (bobs4462) [reply src/pubsub.rs]: Agree, changed it to `Pubkey`, also rebased the branch on top of master

File: src/rpc.rs [hunk 2]
```diff
@@ -517,13 +543,26 @@ impl Cacheable for GetAccountInfo {
         state.account_info_request_limit.as_ref()
     }
 
+    // for getAccountInfo requests, we don't need to subscribe in case if the owner program exists,
+    // and there's already an active subscription present for it
+    fn has_active_subscription(&self, state: &State) -> SubscriptionActive {
+        let owner = state.accounts.get(&self.pubkey).and_then(|data| {
```
Comment (polachok) [inline src/rpc.rs:549]: Why is there an extra map lookup for owner if we have this data already?

Comment (bobs4462) [reply src/rpc.rs:549]: Hmm, I cannot seem to find that data, we have cached response, yes, but it's serialiazed, may be i'm missing something here...

Comment (polachok) [reply src/rpc.rs:549]: I suppose we can return owner from get_from_cache

Comment (bobs4462) [reply src/rpc.rs:549]: Yes, we can, but it's not really easy to make use of this value
for example https://github.com/bobs4462/cacherpc/blob/sub-deduplication/src/rpc.rs#L363 here we can use that value and pass it to `has_active_subscription`, but https://github.com/bobs4462/cacherpc/blob/sub-deduplication/src/rpc.rs#L442 in this invocation we don't have access to that value. It's possible to to keep it around, but overall it's going to be pretty messy in my opinion.

Comment (bobs4462) [reply src/rpc.rs]: Owner was added at call site of the `has_active_subscription`, to eliminate extra cache lookup

File: src/pubsub.rs [hunk 3]
```diff
@@ -1035,6 +1084,21 @@ impl StreamHandler<AccountCommand> for AccountUpdateManager {
...
+    type Result = bool;
+
+    fn handle(&mut self, item: IsSubActive, _: &mut Context<Self>) -> bool {
+        let owner_sub = item
+            .owner
+            .map(|sub| {
+                self.sub_to_id
+                    .contains_key(&(Subscription::Program(sub), item.commitment))
+            })
+            .unwrap_or(false);
```
[the remark this answers is not in the document]
Comment (bobs4462) [reply src/pubsub.rs:1097]: Added additional metric for tracking those cases :heavy_check_mark:

[the remark this answers is not in the document]
Comment (bobs4462) [reply src/pubsub.rs]: Also refactored incorrect check for owner subscription

File: src/pubsub.rs [hunk 4]
```diff
@@ -101,6 +104,17 @@ impl Future for SubscriptionActive {
         let mut inner = Pin::into_inner(self);
         match &mut inner {
             SubscriptionActive::Ready(res) => Poll::Ready(*res),
+            SubscriptionActive::RequestWithOwner(ref mut req) => match Pin::new(req).poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready((Ok(owner_sub_exists), Ok(sub_exists))) => {
```
Comment (polachok) [inline src/pubsub.rs:109]: Wouldn't one Ok(_) be enough? It seems like it wouldn't matter if the other is Err(_)

Comment (bobs4462) [reply src/pubsub.rs]: Done

File: src/pubsub.rs [hunk 5]
```diff
@@ -1294,6 +1379,14 @@ pub enum AccountCommand {
     Purge(Subscription, Commitment),
 }
 
+#[derive(Message, Debug)]
+#[rtype(result = "bool")]
+pub struct IsSubActive {
+    sub: Subscription,
+    commitment: Commitment,
+    //owner: Option<Pubkey>,
```
Comment (polachok) [inline src/pubsub.rs:1387]: ?

Comment (bobs4462) [reply src/pubsub.rs]: fixed, forgot to remove the last time

**Review by polachok — APPROVED** [review]

**Review by polachok — APPROVED** [review]

**Pull request merged**